### PR TITLE
generic webhook: set a default timeout for webhook requests

### LINKED ISSUE #53698: generic webhook restClient should set http timeout
**Is this a BUG REPORT or FEATURE REQUEST?**:
> /kind bug

**What happened**:
when network exception, webhook for authentication/audit/admission can be blocked out of control.
**What you expected to happen**:
Should make all http request under control
**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration**:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by ericchiang**
Add a 30 second timeout for all HTTP requests that the webhook sends
so they timeout instead of hanging forever.

closes https://github.com/kubernetes/kubernetes/issues/53698

cc @kubernetes/sig-api-machinery-pr-reviews 

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go [hunk 1]
```diff
@@ -32,13 +32,21 @@ import (
...
+const defaultRequestTimeout = 20 * time.Second
+
 type GenericWebhook struct {
 	RestClient     *rest.RESTClient
 	initialBackoff time.Duration
 }
 
 // NewGenericWebhook creates a new GenericWebhook from the provided kubeconfig file.
 func NewGenericWebhook(registry *registered.APIRegistrationManager, codecFactory serializer.CodecFactory, kubeConfigFile string, groupVersions []schema.GroupVersion, initialBackoff time.Duration) (*GenericWebhook, error) {
+	return newGenericWebhook(registry, codecFactory, kubeConfigFile, groupVersions, initialBackoff, defaultRequestTimeout)
```
Comment (hzxuzhonghu) [inline staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go]: Should make this configurable?

Comment (ericchiang) [reply staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go]: Configurable by who? You mean adding new API flags or adding this value to the admission webhook API?

File: staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go [hunk 2]
```diff
@@ -53,6 +61,14 @@ func NewGenericWebhook(registry *registered.APIRegistrationManager, codecFactory
 	if err != nil {
 		return nil, err
 	}
+
+	// Kubeconfigs can't set a timeout, this can only be set through a command line flag.
+	//
+	// https://github.com/kubernetes/client-go/blob/master/tools/clientcmd/overrides.go
+	//
+	// Set this to something reasonable so request to webhooks don't hang forever.
+	clientConfig.Timeout = requestTimeout
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go]: @roycaihw could you confirm if this is the right way to set client timeout after your recent refactoring?

File: staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go [hunk 3]
```diff
@@ -32,13 +32,21 @@ import (
 	"k8s.io/client-go/tools/clientcmd"
 )
 
+// defaultRequestTimeout is set for all webhook request. This is the absolute
+// timeout of the HTTP request, including reading the response body.
+const defaultRequestTimeout = 20 * time.Second
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go]: nit: For consistency with the [admission webhook](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/webhook/admission.go#L284), could you set it to 30s?

File: staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go (hunk 1 above)
Comment (hzxuzhonghu) [reply staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go]: I am not sure ,  may be this 
>// TODO(ericchiang): Make these value configurable. Maybe through a
>	// kubeconfig extension?

File: staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go (hunk 2 above)
Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go]: nvm, I think it's still the correct API.

**Comment by caesarxuchao** [comment]
lgtm except for a nit.

File: staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go (hunk 3 above)
Comment (ericchiang) [reply staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go]: done

File: staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go (hunk 1 above)
Comment (ericchiang) [reply staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go]: If we're not sure let's just leave it as is. If there's a strong reason to control this timeout we can argue for that later.

**Comment by caesarxuchao** [comment]
I think it's fine hard coding the timeout. We can add a flag later if necessary.

/lgtm

**Pull request merged**